File: test/interop/python/basic_client.py
import argparse
import asyncio

from mcp import ClientSession, StdioServerParameters
from mcp.client.stdio import stdio_client


async def run(
    command: str,
    server_args: list[str],
    cwd: str | None,
    expect_inspector_primitives: bool,
) -> None:
    params = StdioServerParameters(command=command, args=server_args, cwd=cwd)
    async with stdio_client(params) as (read, write):
        async with ClientSession(read, write) as session:
            await session.initialize()
            tools = await session.list_tools()
            names = {tool.name for tool in tools.tools}
            if "echo" not in names:
                raise RuntimeError(f"Expected echo tool, got {sorted(names)}")

            result = await session.call_tool(
                "echo",
                {"message": "from python client"},
            )
            content = result.content[0]
            text = getattr(content, "text", None)
            if text != "from python client":
                raise RuntimeError(f"Unexpected echo result: {result!r}")

            if expect_inspector_primitives:
                resources = await session.list_resources()
                uris = {str(resource.uri) for resource in resources.resources}
                if "inspector://status" not in uris:
                    raise RuntimeError(f"Expected inspector resource, got {sorted(uris)}")

                prompts = await session.list_prompts()
                names = {prompt.name for prompt in prompts.prompts}
                if "inspector-summary" not in names:
                    raise RuntimeError(f"Expected inspector prompt, got {sorted(names)}")


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--server-command", required=True)
    parser.add_argument("--server-args", default="")
    parser.add_argument("--server-cwd")
    parser.add_argument("--expect-inspector-primitives", action="store_true")
    args = parser.parse_args()
    server_args = [part for part in args.server_args.split(" ") if part]
    asyncio.run(
        run(
            args.server_command,
            server_args,
            args.server_cwd,
            args.expect_inspector_primitives,
        )
    )
    print("python client interop passed")


if __name__ == "__main__":
    main()

File: test/interop/ts/src/basic_client.ts
import { Client } from '@modelcontextprotocol/sdk/client/index.js';
import { StdioClientTransport } from '@modelcontextprotocol/sdk/client/stdio.js';
import {
  CreateMessageRequestSchema,
  ElicitRequestSchema,
  ListRootsRequestSchema,
} from '@modelcontextprotocol/sdk/types.js';

function readArg(args: string[], name: string): string | undefined {
  const index = args.indexOf(name);
  if (index < 0 || index + 1 >= args.length) {
    return undefined;
  }
  return args[index + 1];
}

function readServerArgs(args: string[]): string[] {
  const value = readArg(args, '--server-args');
  return value ? value.split(' ').filter((part) => part.length > 0) : [];
}

function hasFlag(args: string[], name: string): boolean {
  return args.includes(name);
}

async function main(): Promise<void> {
  const args = process.argv.slice(2);
  const command = readArg(args, '--server-command');
  const activeClientCapabilities = hasFlag(
    args,
    '--active-client-capabilities'
  );
  if (!command) {
    throw new Error('--server-command is required');
  }

  const transport = new StdioClientTransport({
    command,
    args: readServerArgs(args),
    cwd: readArg(args, '--server-cwd'),
  });
  const client = new Client(
    { name: 'ts-basic-client', version: '1.0.0' },
    {
      capabilities: activeClientCapabilities
        ? {
            roots: { listChanged: true },
            sampling: {},
            elicitation: { form: {} },
          }
        : {},
    }
  );

  if (activeClientCapabilities) {
    client.setRequestHandler(ListRootsRequestSchema, async () => ({
      roots: [{ uri: 'file:///tmp/mcp-inspector', name: 'Inspector Root' }],
    }));
    client.setRequestHandler(CreateMessageRequestSchema, async () => ({
      model: 'inspector-mock-model',
      role: 'assistant' as const,
      content: {
        type: 'text' as const,
        text: 'inspector sampling response',
      },
    }));
    client.setRequestHandler(ElicitRequestSchema, async () => ({
      action: 'accept' as const,
      content: { confirmed: true },
    }));
  }

  try {
    await client.connect(transport);

    const tools = await client.listTools();
    const toolNames = tools.tools.map((tool) => tool.name);
    if (!toolNames.includes('echo')) {
      throw new Error(`Expected echo tool, got ${toolNames.join(', ')}`);
    }

    const result = await client.callTool({
      name: 'echo',
      arguments: { message: 'from typescript client' },
    });
    const content = result.content as Array<{ type: string; text?: string }>;
    const text = content[0];
    if (text.type !== 'text' || text.text !== 'from typescript client') {
      throw new Error(`Unexpected echo result: ${JSON.stringify(result)}`);
    }

    if (hasFlag(args, '--expect-inspector-primitives')) {
      const resources = await client.listResources();
      const resourceUris = resources.resources.map((resource) => resource.uri);
      if (!resourceUris.includes('inspector://status')) {
        throw new Error(
          `Expected inspector://status resource, got ${resourceUris.join(', ')}`
        );
      }

      const prompts = await client.listPrompts();
      const promptNames = prompts.prompts.map((prompt) => prompt.name);
      if (!promptNames.includes('inspector-summary')) {
        throw new Error(
          `Expected inspector-summary prompt, got ${promptNames.join(', ')}`
        );
      }
    }

    console.log('typescript client interop passed');
  } finally {
    await client.close();
  }
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});
